crates/nostr-sqldb/src/postgres.rs

```diff
@@ -0,0 +1,235 @@
+                }
+                e => Err(DatabaseError::backend(e)),
+            },
+        }
+    }
+
+    pub(crate) async fn event_by_id(
+        &self,
+        event_id: &EventId,
+    ) -> Result<Option<EventDb>, DatabaseError> {
+        let res = event_by_id(event_id)
+            .first(&mut self.get_connection().await?)
+            .await
+            .optional()
+            .map_err(DatabaseError::backend)?;
+        Ok(res)
+    }
+}
+
+impl NostrEventsDatabase for NostrPostgres {
+    /// Save [`Event`] into store
+    ///
+    /// **This method assumes that [`Event`] was already verified**
+    fn save_event<'a>(
+        &'a self,
+        event: &'a Event,
+    ) -> BoxedFuture<'a, Result<SaveEventStatus, DatabaseError>> {
+        Box::pin(async move { self.save(EventDataDb::try_from(event)?).await })
+    }
+
+    /// Check event status by ID
+    ///
+    /// Check if the event is saved, deleted or not existent.
+    fn check_id<'a>(
+        &'a self,
+        event_id: &'a EventId,
+    ) -> BoxedFuture<'a, Result<DatabaseEventStatus, DatabaseError>> {
+        Box::pin(async move {
+            let status = match self.event_by_id(event_id).await? {
+                Some(e) if e.deleted => DatabaseEventStatus::Deleted,
+                Some(_) => DatabaseEventStatus::Saved,
+                None => DatabaseEventStatus::NotExistent,
+            };
+            Ok(status)
+        })
+    }
+
+    /// Coordinate feature is not supported yet
+    fn has_coordinate_been_deleted<'a>(
+        &'a self,
+        _coordinate: &'a nostr::nips::nip01::CoordinateBorrow<'a>,
+        _timestamp: &'a Timestamp,
+    ) -> BoxedFuture<'a, Result<bool, DatabaseError>> {
+        Box::pin(async move { Ok(false) })
+    }
+
+    /// Get [`Event`] by [`EventId`]
+    fn event_by_id<'a>(
+        &'a self,
+        event_id: &'a EventId,
+    ) -> BoxedFuture<'a, Result<Option<Event>, DatabaseError>> {
+        Box::pin(async move {
+            let event = match self.event_by_id(event_id).await? {
+                Some(e) if !e.deleted => {
+                    Some(Event::decode(&e.payload).map_err(DatabaseError::backend)?)
+                }
+                _ => None,
+            };
+            Ok(event)
+        })
+    }
+
+    /// Count the number of events found with [`Filter`].
+    ///
+    /// Use `Filter::new()` or `Filter::default()` to count all events.
+    fn count(&self, filter: Filter) -> BoxedFuture<Result<usize, DatabaseError>> {
+        Box::pin(async move {
+            let res: i64 = build_filter_query(filter)
+                .count()
+                .get_result(&mut self.get_connection().await?)
+                .await
+                .map_err(DatabaseError::backend)?;
+            Ok(res as usize)
+        })
+    }
+
+    /// Query stored events.
+    fn query(&self, filter: Filter) -> BoxedFuture<Result<Events, DatabaseError>> {
```

Comment:
It would be nice if we could return a async iterator here and let nostr-relay-builder consume it without allocating all the events in memory.